### PR TITLE
Hostname SPI v2

### LINKED ISSUE #26084: Implement Hostname SPI v2
### Description

The implementation should follow the design specified in https://github.com/keycloak/keycloak/issues/24252.

### PR DESCRIPTION
**Pull request by vmuzikar**
Closes: #26084

### CODE REVIEW

File: common/src/main/java/org/keycloak/common/profile/PropertiesProfileConfigResolver.java
Comment (shawkins) [inline common/src/main/java/org/keycloak/common/profile/PropertiesProfileConfigResolver.java:49]: the parameter name for Profile.getFeatureConfig needs updated - it implies that only unversioned is being passed, but that's not the case.

Comment (vmuzikar) [reply common/src/main/java/org/keycloak/common/profile/PropertiesProfileConfigResolver.java:49]: Good point, I'll fix that. Thanks.

Comment (vmuzikar) [reply common/src/main/java/org/keycloak/common/profile/PropertiesProfileConfigResolver.java:49]: Done.

**Comment by vmuzikar** [comment]
@keycloak/cloud-native This PR is currently mostly complete. It includes:
* The actual Hostname SPI implementation which now lives in the core Keycloak codebase as it doesn't depend on Quarkus in any way anymore.
* `hostname:v2` profile feature.
* Integration tests.
* Tweaks to `hostname-debug`.

What's still missing is:
* CLI options (blocked by #25333).
* Quarkus integration tests (blocked by #25333).
* Docs (will be done as a separate commit by @Pepo48), or as a follow-up PR.

Since the current content of this PR is expected to remain as is (the missing stuff will be just added on top of it), it should be ready for an initial review.

**Comment by mabartos** [comment]
@vmuzikar How exactly do you want to address options sections? Do you plan to introduce some new OptionCategory, with different `ConfigSupportLevel`? I mean f.e HOSTNAME_V2 with PREVIEW ConfigSupportLevel. 

I think there are 2 options:

1. Introduce a new option category: The sections in the help would be separated (we could deprecate the whole category). When a specific version of the feature is disabled, the whole category is hidden. For `help-all`, categories are separated and UX is better.
2. Do NOT introduce a new option category: For `help-all`, all options for the category are printed out and merged together - not great UX.

_Could be also related to the option duplicability, but I'm still in the process of addressing used use cases._

**Comment by vmuzikar** [comment]
@mabartos You mean a category like "Hostname V2"? Now we have just "Hostname". Seems alright to me, we could then remove the "V2" suffix once we finally remove the old options.

**Comment by mabartos** [comment]
> @mabartos You mean a category like "Hostname V2"? Now we have just "Hostname". Seems alright to me, we could then remove the "V2" suffix once we finally remove the old options.

Yes, I do. I think it'd be a better approach to provide the new category and once we want to remove the old options, we remove the options in the "Hostname" category and remove the suffix from the "Hostname V2" (replace them :).

**Comment by vmuzikar** [comment]
So we'd display only one of the categories (either "Hostname" or "Hostname V2") in `help` based on enabled profile features, but display both in `help-all` and docs? Makes sense to me. 👍

**Comment by mabartos** [comment]
> So we'd display only one of the categories (either "Hostname" or "Hostname V2") in help based on enabled profile features, but display both in help-all and docs? Makes sense to me. 👍

@vmuzikar Exactly :)

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/HostnameV1PropertyMappers.java
Comment (mabartos) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/HostnameV1PropertyMappers.java:44]: +1

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMappers.java
Comment (mabartos) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMappers.java:65]: @vmuzikar I'm afraid it's not gonna work. As for the proper determination of disabled mappers, the config sources need to be initialized. AFAIK, when this is done before the cli.parseArgs[1], the `--config-file` property will not work as you'd need to reinit particular config sources after evaluating that. That was the main reason to delay the sanitization. If I remember correctly, it had some impact also on the Profile initialization, but not sure. But I believe it could be somehow "workarounded".

[1] https://github.com/keycloak/keycloak/pull/26345/files#diff-5e7526010bd1f3d3c2735d6b5d600aaf3457a83eb1699ef7e10c91e2257ae6d2L113

Comment (vmuzikar) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMappers.java]: Thanks for the pointer! You are correct, I implemented it in a different way.

**Comment by vmuzikar** [comment]
This PR should be now ready to review. It includes:
* Hostname v2 SPI implementation.
* **Changes that makes v2 the default and deprecates v1.** It is possible to switch back to Hostname v1 via `--features hostname:v1`.**
* New config options. Those are separated in a "Hostname v2" category. Old Hostname v1 remains in its own category. Each category is either enabled or disabled based on `--features hostname:v1|2`. This is also reflected in `--help` while `--help-all` displays both categories.
* Adjustments for the logic around conditional config options to support duplicit option names. It's a bit of a hack but it works nicely. This is an important piece as e.g. `--hostname` option is a different option for v1 and v2.
* Removal of some old and no longer relevant hostname implementations. Those were excluded from the dist and were used only by the testsuite on Undertow (as current Hostname v1 works only for Quarkus dist). Hostname v2 does not depend on Quarkus, so it can be used by Undertow as well.
* Testsuite tweaks, mostly related to making enablement and disablement of features work with Quarkus. This was needed by the original version of the PR where v2 was not the default. It is no longer strictly needed but might still be useful in the future if a test requires this. I can of course remove those changes if needed.

Since this PR is already quite big, the following will be resolved as follow-ups:
* https://github.com/keycloak/keycloak/issues/27729
* https://github.com/keycloak/keycloak/issues/27728
* https://github.com/keycloak/keycloak/issues/27730

CC @keycloak/cloud-native @keycloak/maintainers @keycloak/maintainers-external

**Comment by vmuzikar** [comment]
jFTR, this is how it looks in docs:
Screenshot 2024-03-20 at 16 09 18

**Review by mabartos — COMMENTED** [review]
@vmuzikar Very nice job! I'll try to review it tomorrow. I just spotted one minor typo.

**Review by abstractj — DISMISSED** [review]
The proposed changes LGTM, and the test failures seem unrelated. Therefore, I'm approving it.

**Comment by shawkins** [comment]
LGTM, but HostnameV2DistTest.testServerFailsToStartWithoutHostnameSpecified is failing.

**Comment by vmuzikar** [comment]
@abstractj @shawkins Thanks for the review! I fixed `HostnameV2DistTest`. It was caused by the rebase, root cause were probably changes introduced in https://github.com/keycloak/keycloak/pull/28225.

**Review by mabartos — COMMENTED** [review]
@vmuzikar Very nice job! In general LGTM. 

I haven't reviewed the Hostname V2 provider impl yet, but I put some nitpick comments below. Not something that could block this PR anyway.

**Comment by mabartos** [comment]
@vmuzikar Other than that, LGTM

**Review by mabartos — DISMISSED** [review]
As I've mentioned, these nitpicks above are not meant to block this PR. LGTM!

**Review by abstractj — DISMISSED** [review]

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMappers.java
Comment (vmuzikar) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMappers.java]: Good catch! For `getMapperOrDefault` I just kept the previous behavior, i.e. using unpolished. It's not a new method. I went ahead and changed it but not sure if this behaviour wasn't intentional. Let's see if tests pass.

File: quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/HostnameV2DistTest.java
Comment (vmuzikar) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/HostnameV2DistTest.java:43]: That's a good point and I was actually thinking about it when writing it. I'd still like to keep it here. Other tests might be testing this as well but only as a "side effect" and we could easily forget about that when refactoring the tests for example. This test explicitly covers this functionality so it's more obvious.

File: services/src/main/java/org/keycloak/url/HostnameV2ProviderFactory.java
Comment (vmuzikar) [reply services/src/main/java/org/keycloak/url/HostnameV2ProviderFactory.java]: Done, thanks.

File: quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/HostnameV2DistTest.java
Comment (mabartos) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/HostnameV2DistTest.java:43]: ok, works for me

**Review by mabartos — DISMISSED** [review]
@vmuzikar Thanks for these changes, very nice job!

**Review by Pepo48 — DISMISSED** [review]
I went through all the recent changes and it LGTM. Thanks, @vmuzikar; very nicely done!

**Comment by shawkins** [comment]
@vmuzikar some of the tests seem to be failing with an underlying hostname error:

```
2024-04-04T12:09:27.3021209Z ImportAtStartupDistTest ++ ERROR: hostname is not configured; either configure hostname, or set hostname-strict to false
```

This seems to happening during an export operation - so we're probably over validating during import / export.

**Comment by vmuzikar** [comment]
@shawkins @mabartos I had to revert the fix around polishing the property names. There seem to be something else going on and it doesn't seem to be simple to fix. I will look into it more and file a separate PR if necessary.

**Review by mabartos — DISMISSED** [review]
@vmuzikar LGTM.

> I will look into it more and file a separate PR if necessary.

It'd be good as I'm kinda curious about what's happening there.

**Review by shawkins — DISMISSED** [review]
LGTM as well, thanks @vmuzikar

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMappers.java
Comment (mabartos) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMappers.java:159]: > I did not find any reason to keep the filtering. If more than 1 mapper was found for an option, we don't expect one would be in an enabled category while others in disabled. So I removed the filtering.

It was more like an enhancement/extension to the property mapper search. When the current command does not contain(`AbstractCommand.getOptionCategories()`) the mapper's category, it's just ignored.  

@vmuzikar But, the current approach works for me. Thanks for the more elaborate explanation.

**Review by mabartos — APPROVED** [review]
LGTM

**Comment by vmuzikar** [comment]
@mabartos Thanks for yet another re-review. :)

@shawkins @abstractj Does it look good to you too?

**Review by shawkins — APPROVED** [review]
LGTM

**Review by abstractj — APPROVED** [review]

**Review by abstractj — APPROVED** [review]

**Comment by vmuzikar** [comment]
jFTR since this is a breaking change, docs are handled in a separate PR: https://github.com/keycloak/keycloak/pull/28123

**Pull request merged**
